ibis/expr/builders.py

```diff
@@ -53,19 +52,14 @@ def when(self, case_expr: Any, result_expr: Any) -> Self:
         result_expr
             Value when the case predicate evaluates to true.
         """
-        return self.copy(
-            cases=self.cases + (case_expr,), results=self.results + (result_expr,)
-        )
+        self.cases += (case_expr,)
```

Comment:
The expression builders are now mutable objects inheriting from `Annotable` rather than from `Concrete`. This was necessary to support the unhashable `Deferred` objects.

Reply:
Switched backed to be immutable because `_.col.sum().over(frame)` should be supported where frame is a window builder and the entire object should be hashable.